### ISSUE TITLE
Sort all branches by last update

### ISSUE THREAD

**Issue by hubgit** [desktop/desktop#5155]
**Is your feature request related to a problem? Please describe.**

Showing recent branches is a great feature, but I often have trouble finding a fairly recent branch because it's disappeared from the short "Recent Branches" list.

**Describe the solution you'd like**

Divide the list of branches into "Local Branches" (sorted by date) and "Remote Branches" (sorted by date), instead of "Recent Branches" (sorted by date) and "Other Branches" (sorted alphabetically).

It's not very useful to have branches sorted alphabetically, as you can already type to filter them.

**Describe alternatives you've considered**

Switching to the command line and using a custom `git branches` alias:

```bash
for-each-ref --sort=committerdate refs/heads/ --format='%(HEAD) %(color:yellow)%(refname:short)%(color:reset) - %(color:red)%(objectname:short)%(color:reset) - %(contents:subject) - %(authorname) (%(color:green)%(committerdate:relative)%(color:reset))'
```

**Teachability, Documentation, Adoption, Migration Strategy**

This would replace the current dropdown list of branches.

**Comment by shiftkey** [comment]
> Showing recent branches is a great feature, but I often have trouble finding a fairly recent branch because it's disappeared from the short "Recent Branches" list.

I'm actually more interested in this part of the issue than changing the sorting, because I've spent a bit of time in there thinking about how it works.

I think the core of your problem is that **we don't consider whether a branch exists in the list when we're reading the reflog**. I can see this in my local Desktop clone after pruning some merge branches:

<img width="398" src="https://user-images.githubusercontent.com/359239/42841842-ba5762d6-89e1-11e8-80cd-5f76e512fecd.png">

Are you pruning your merged branches? I think the fact that our `getRecentBranches` function here doesn't confirm a branch exists locally mean we can end up often showing fewer branches than the 5 allowed.

https://github.com/desktop/desktop/blob/56b1fe71f59d3f59202f5949ebc48af606e5f89c/app/src/lib/git/reflog.ts#L5-L8

**Comment by hubgit** [comment]
> Are you pruning your merged branches? I think the fact that our getRecentBranches function here doesn't confirm a branch exists locally mean we can end up often showing fewer branches than the 5 allowed.

The remote branches are being deleted when they're merged, but the local branches aren't being pruned. Does the list take account of which branches are in the origin repository and which have been merged, or is it literally just the most recently-edited local branches?

If I run the `git branches` alias that's in the issue description, I can see 5 local branches edited in the last 24hr, but GitHub Desktop only shows 4 of them (+ master). 

There could easily be more than 5 branches edited in the last 24hr, as well.

I'm interested to know when it's useful to have branches listed in alphabetical order, rather than by date.

**Comment by shiftkey** [comment]
@hubgit thanks for the extra details

> Does the list take account of which branches are in the origin repository and which have been merged, or is it literally just the most recently-edited local branches?

No, the code has zero context on the actual branch details and just walks the reflog to find the N most recent branches that were checked out locally. This is what we consider "recent" - branches that you've checked out locally, and we don't look at the most recent commit to each branch.

> I can see 5 local branches edited in the last 24hr, but GitHub Desktop only shows 4 of them (+ master).

This makes sense - we mark `master` as the default branch, so it gets listed before these recent branches and we show one fewer.

> I'm interested to know when it's useful to have branches listed in alphabetical order, rather than by date.

This is mostly just mirroring the defaults of `git branch`

**Comment by oguzyildiz1991** [comment]
It would be also great if I could increase the number of recent branches displayed in there. Once a branch I'm trying to get is 6th most recent one, finding it takes a bit time

**Comment by billygriffin** [comment]
Thanks @oguzyildiz1991! Because we've heard that most people are working between 2-4 branches, we decided to restrict the number to be fairly low to provide that ability to more easily switch back and forth between a few, but also not cloud the larger list for folks who prefer navigating to repos in a different way. It's tricky to balance everyone's desires here, but for now we don't have plans to expand the recent list.

**Comment by outofambit** [comment]
Thanks for the bump here @oguzyildiz1991 ! There's another conversation happening in #7755 that I've linked this issue to.

**Comment by jcushman** [comment]
I'll just add another user story for this feature. I often use "recent branches" to switch between the last few branches when I'm working on a couple of things at once. But if I start working on something and then move away for a little longer, I'll have to do something like `git for-each-ref --sort=-committerdate refs/heads/ | less` to find it again because the list only has five items. "Other branches" has hundreds of branches and doesn't serve any use for me when ordered alphabetically.

**Comment by oguzyildizz** [comment]
Exactly. If you're working on 5 branches it's alright, but once you're looking for the 6th most recent branch, suddenly you gotta go on an adventure

**Comment by steveward** [comment]
In https://github.com/desktop/desktop/issues/13105 it was brought up that remote branches do not appear in the `Recent branches` list. Showing them would align with what was mentioned in https://github.com/desktop/desktop/issues/5155#issuecomment-407794598:

>but if you're in a more collaborative situation using the commit timestamp means you might able to keep up with activity from collaborators

**Comment by teadrinker** [comment]
The way I perceive recent, is "locally recent" - which branches I have actually recently used/switched to, on this machine).
I think this functionality is fine, and I don't need more than 3 of these for my own usage. 

However, it is very annoying that I cannot see new branches that have been added remotely (we have a lot of branches).
For this issue, I would suggest a selection of sorting (Name or Date), ideally this option is global, and saved automatically with app settings.

![dropdown](https://user-images.githubusercontent.com/678065/158080269-4008164a-1941-4f0b-92b3-a4f46222d605.jpg)

**Comment by Janpot** [comment]
Another option could be to add a "show more..." button at the bottom of the 5 most recent branches, to increase the list with 5 more.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/14828

**Comment by steveward** [comment]
From https://github.com/desktop/desktop/issues/16897:

>When attempting to switch to a new branch, the Recent Branches list typically does not show the actual most recent branch(es). If I don't know the branch name, I have to open GitHub on the web, where the Active Branches list does show the most recent branches.

**Comment by steveward** [comment]
From https://github.com/desktop/desktop/issues/19487:

>This would be helpful when working with other developers or switching machines. The filter works to find the branch given you know the name, however always having the most recently changed branch at the top would make ux feel faster a I don't need to see stale branches at the top of the list.

**Comment by adamreisnz** [comment]
It would be great if the "Other branches" could show them in order of last commit, as the [Docs](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/viewing-branches-in-your-repository) seem to imply is happening.

But in my experience this is not happening at all, with branches that have had no commits for 7 months showing first, above branches with commits a few days ago:
![image](https://github.com/user-attachments/assets/4b7d5f96-46d6-4418-b948-b9826501a819)

This is making it very hard to find recent branches if you are working on more than 5 branches regularly.

(Am aware that some old branches need to be cleaned up and removed, unfortunately it has not been a priority yet)

**Comment by trumad** [comment]
I would also find it helpful if these could be ordered by date. Or give the user a choice if they'd rather sort alphabetically.

**Comment by tidy-dev** [comment]
From #21132 

> Would be great if we ha a visual information (a red background or any color else) of the last branch commited by date.
> Add a color in the background of the last item commited by date.

**Comment by msawired** [comment]
> The way I perceive recent, is "locally recent" - which branches I have actually recently used/switched to, on this machine). I think this functionality is fine, and I don't need more than 3 of these for my own usage.
> 
> However, it is very annoying that I cannot see new branches that have been added remotely (we have a lot of branches). For this issue, I would suggest a selection of sorting (Name or Date), ideally this option is global, and saved automatically with app settings.
> 
> ![dropdown](https://user-images.githubusercontent.com/678065/158080269-4008164a-1941-4f0b-92b3-a4f46222d605.jpg)

on another ticket, I also suggested the exact solution. Adding an option to change ordering alphabetical vs. date in Other Brances would solve lots of problems mentioned here: Would help finding "other" recent branches not displayed in "recent branches", and it would also help notice and clean stalled branches.

**Comment by metasidd** [comment]
Love that solution @msawired . I believe this would be helpful even after a user types in a search query. Given that projects can have thousands of branches (open and closed), I'd love to see it sorted by last updated.

**Comment by greegus** [comment]
This feature is so needed – once you work with more than a few branches that you checkout between often, it's practically impossible with this app to get to one without knowing an exact name.

The sorting of the unfiltered (no search query) list of branched just doesn't make ANY sense to me. WHY would anybody want to see a 2-year-old branch in front of a 3-month-old one, followed by a 9-month-old one, when trying to check out last week's work?! It's not even alphabetically sorted. Literally counterproductive.

**Comment by adesai1000** [comment]
For anyone still waiting on this since 2018: my fork, GlitchHub Desktop, adds it. Settings > Appearance > Branch List has "Sort other branches by" with a "Last update, newest first" option. The Default and Recent groups keep working the way they do today; the long tail of other branches is what gets sorted by the tip commit date. You can also right-click any branch to pin it to the top of the list.

Not affiliated with GitHub. macOS on Apple silicon only for now, unsigned build, so right-click and Open the first time. Separate profile folder, so it lives happily next to GitHub Desktop.

Download: https://github.com/adesai1000/GlitchHub-Desktop/releases/latest

If it is useful, a star on the repo would help other people here find it.
